kitchen.yml

```
@@ -429,64 +429,17 @@ verifier:
 
 suites:
   - name: default
-    excludes:
-      - gentoo-stage3-latest-master-py3
-      - gentoo-stage3-systemd-master-py3
-      - gentoo-stage3-latest-3003-1-py3
-      - gentoo-stage3-systemd-3003-1-py3
-      - gentoo-stage3-latest-3002-6-py3
-      - gentoo-stage3-systemd-3002-6-py3
-      - gentoo-stage3-latest-3001-7-py3
-      - gentoo-stage3-systemd-3001-7-py3
-      - gentoo-stage3-latest-3000-9-py3
-      - gentoo-stage3-systemd-3000-9-py3
     provisioner:
       state_top:
         base:
           '*':
             - TEMPLATE._mapdata
             - TEMPLATE
-      pillars:
-        top.sls:
-          base:
-            '*':
-              - TEMPLATE
-              - define_roles
+      pillars: {}
       pillars_from_files:
         TEMPLATE.sls: pillar.example
-        define_roles.sls: test/salt/pillar/define_roles.sls
-    verifier:
-      inspec_tests:
-        - path: test/integration/default
-  - name: gentoo
-    includes:
-      - gentoo-stage3-latest-master-py3
-      - gentoo-stage3-systemd-master-py3
-      - gentoo-stage3-latest-3003-1-py3
-      - gentoo-stage3-systemd-3003-1-py3
-      - gentoo-stage3-latest-3002-6-py3
-      - gentoo-stage3-systemd-3002-6-py3
-      - gentoo-stage3-latest-3001-7-py3
-      - gentoo-stage3-systemd-3001-7-py3
-      - gentoo-stage3-latest-3000-9-py3
-      - gentoo-stage3-systemd-3000-9-py3
-    provisioner:
-      state_top:
-        base:
-          '*':
-            - TEMPLATE._mapdata
-            - TEMPLATE
-      pillars:
-        top.sls:
-          base:
-            '*':
-              - TEMPLATE
-              - gentoo
-              - define_roles
```

We probably can avoid a dedicated `test/salt/pillar/top.sls` and just adapt the same method used in other formulas under `state_top`.

Some examples:

* [logrotate-formula](https://github.com/saltstack-formulas/logrotate-formula/blob/be7d5795ecd625a9e0427c1b81f328144b1c78d7/kitchen.yml#L453-L457)
* [ntp-formula](https://github.com/saltstack-formulas/ntp-formula/blob/878a101f6cedb88885798d9c42d157a11593fcbe/kitchen.yml#L453-L457)
* [rng-tools-formula](https://github.com/saltstack-formulas/rng-tools-formula/blob/0a0d041bcad00669244999c52a4466401a97ee01/kitchen.yml#L453-L457)

```suggestion
      pillars:
        top.sls:
          base:
            '*':
              - TEMPLATE
              - define_roles
            'G@os_family:Gentoo':
                - gentoo
```

---

Yes, It could be done from `kitchen.yml` directly but the main point of having a dedicated `top.sls` is to be able to run it outside kitchen.

---

@baby-gnu I haven't come across this idea before -- could you give me an example of this being run outside of Kitchen?  Does this also apply to the using the `state_top`'s `top.sls` as a separate file as well?

---

The idea came [from user colttt on SaltStack matrix room](https://matrix.to/#/!WmgUPuThcUZvfCoQLZ:matrix.org/$162920593550701UaMAC:matrix.org) to better validate formula since `salt-lint` does not cover all possible errors.

colttt uses the following command to compile the states:

```
salt-call --id=test-minion.example.de --local --file-root=/srv/salt/states --pillar-root=/srv/salt/pillars --retcode-passthrough state.show_sls
```

But it can't be done easily when the formula requires pillars.

Then, my idea is to provide all the necessary to try out a formula in two steps:

1. clone the formula:
   ```
   git clone https://github.com/saltstack-formulas/template-formula.git && cd template-formula

   ```
2. execute the formula locally with our test pillar values:
   ```
   salt-call --local --id test-minion.example.net --file-root=./ --pillar-root=./test/salt/pillar/ state.show_sls TEMPLATE
   ```

There is no point in having `state_top`'s `top.sls` in a dedicated file too since there is no `salt-call` option to select that file and I think we should avoid having one in the formula directory referenced by the `--file-root` option above.

---

@baby-gnu The example you've shared doesn't appear to work, because the `TEMPLATE` pillar (i.e. `pillar.example`) is not in that directory:

```sls
[ERROR   ] Specified SLS 'TEMPLATE' in environment 'base' is not available on the salt master
[ERROR   ] Template was specified incorrectly: False
[CRITICAL] Pillar render error: Specified SLS 'TEMPLATE' in environment 'base' is not available on the salt master
Error running 'state.show_sls': Pillar failed to render. Additional info follows:

- Specified SLS 'TEMPLATE' in environment 'base' is not available on the salt master
```

If I comment out the `TEMPLATE` line in the newly provided `top.sls`, then the state runs -- but obviously missing everything from `pillar.example` (the `TEMPLATE` pillar).

---

Yes, that's an issue.

To be completely functional, the `test/salt/pillar/` should contains all the required files, in which case:
- the `pillars_from_files` is no more used
- the `pillar.example` is just an example and not a seed for tests.

---

I push a second commit to move testing values from `pillar.example` to `test/salt/pillar/TEMPLATE.sls` and remove `pillars_from_files` from `kitchen.yml`.

---

@baby-gnu If we're going to attempt to use these ideas across existing formulas then that's not going to be fun to refactor.  There are more formulas using `pillar.example` in the CI than any other file.  Due to that, it's the default value used from the `ssf-formula` as well.

* [50+ formulas](https://github.com/search?o=desc&q=org%3Asaltstack-formulas+filename%3Akitchen.yml+pillar.example&s=indexed&type=Code) across the whole org.
* Around [37 formulas](https://github.com/search?q=org%3Asaltstack-formulas+filename%3Akitchen.yml+%22pillar.example%22+%22pillars_from_files%22&type=Code) that are managed by the `ssf-formula`.

We can discuss the relative merits of this idea in the upcoming meeting.  I'm still not quite sure what the use cases are so it would great if you could demonstrate it.